### PR TITLE
Add Command Line for Embedding Generating

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

Add command line interface for user to generate embeddings directly.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/run/gs_gen_embedding.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+"""
```
Comment (zheng-da) [inline python/graphstorm/run/gs_gen_embedding.py:1]: can we call it `gs_gen_node_embedding.py`?

Comment (jalencato) [reply python/graphstorm/run/gs_gen_embedding.py]: Yep. I have changed the naming and will leave the folder name still gsgnn_emb, where we can put gsgnn_node_emb.py here, in the future we can put gsgnn_edge_emb.py and gsgnn_graph_emb.py here as well.

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                  use_wholegraph=use_wholegraph(config.part_config))
+    rt_profiler.init(config.profile_path, rank=gs.get_rank())
+    sys_tracker.init(config.verbose, rank=gs.get_rank())
+    device = setup_device(config.local_rank)
+    tracker = gs.create_builtin_task_tracker(config)
+    if gs.get_rank() == 0:
+        tracker.log_params(config.__dict__)
+
+    if config.task_type == BUILTIN_TASK_LINK_PREDICTION:
+        train_data = GSgnnLPTrainData(config.graph_name,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:47]: Shouldn't it GSgnnEdgeInferData? as in lp_infer_gnn.py?

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                                      config.part_config,
+                                      train_etypes=config.train_etype,
+                                      eval_etypes=config.eval_etype,
+                                      node_feat_field=config.node_feat_name,
+                                      pos_graph_feat_field=config.lp_edge_weight_for_loss)
+    elif config.task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
+        train_data = GSgnnNodeTrainData(config.graph_name,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:54]: Shouldn't it GSgnnNodeInferData? as in np_infer_gnn.py?

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                                        config.part_config,
+                                        train_ntypes=config.target_ntype,
+                                        eval_ntypes=config.eval_target_ntype,
+                                        node_feat_field=config.node_feat_name,
+                                        label_field=config.label_field)
+    elif config.task_type in {BUILTIN_TASK_EDGE_CLASSIFICATION, BUILTIN_TASK_EDGE_REGRESSION}:
+        train_data = GSgnnEdgeTrainData(config.graph_name,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:61]: shouldn't it GSgnnEdgeInferData as in ep_infer_gnn?

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    # TODO(zhengda) the model path has to be in a shared filesystem.
+    model.restore_model(model_path)
+    # Preparing input layer for training or inference.
+    # The input layer can pre-compute node features in the preparing step if needed.
+    # For example pre-compute all BERT embeddings
+    model.prepare_input_encoder(train_data)
+    # Runjie: To generate embeddings, it might be preferable to utilize the
+    # entire graph instead of just the training graph.
+    # Additionally, generating embeddings will not result in any edge leakage issues.
+    embeddings = do_full_graph_inference(model, train_data, fanout=config.eval_fanout,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:93]: In some cases, we can only do mini-batch inference.
Please check the implementation of different tasks.

File: python/graphstorm/run/gsgnn_ep/gsgnn_ep.py [hunk 6]
```diff
@@ -156,7 +156,8 @@ def main(config_args):
         save_embeddings(config.save_embed_path, embs, gs.get_rank(),
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_ep/gsgnn_ep.py:156]: You also need to change all the places where save_embeddings is called.

File: python/graphstorm/inference/emb_infer.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        # deal with uninitialized case first
+        if use_mini_batch_infer and \
+                task_type in {BUILTIN_TASK_EDGE_CLASSIFICATION, BUILTIN_TASK_EDGE_REGRESSION}:
+            assert save_embed_path is None, \
+                "Unable to save the node embeddings when using mini batch inference " \
+                "when doing edge task." \
+                "It is not guaranteed that mini-batch prediction will cover all the nodes."
+
+        if task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
+            assert len(loader.data.eval_ntypes) == 1, \
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:82]: GraphStorm supports multiple target node types.

File: python/graphstorm/inference/emb_infer.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                "GraphStorm only support single target node type for training and inference"
+
+        assert save_embed_path is not None
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:85]: Please add an error message

File: python/graphstorm/inference/emb_infer.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                embs = do_mini_batch_inference(self._model, data, fanout=loader.fanout,
+                                               edge_mask=None,
+                                               task_tracker=self.task_tracker)
+            else:
+                embs = do_full_graph_inference(self._model, data, fanout=loader.fanout,
+                                               edge_mask=None,
+                                               task_tracker=self.task_tracker)
+        elif task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
+            # only generate embeddings on the target node type
+            ntype = loader.data.eval_ntypes[0]
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:102]: For ntype in loader.data.eval_ntypes:

File: python/graphstorm/inference/emb_infer.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        node_id_mapping_file: str
+            Path to the file storing node id mapping generated by the
+            graph partition algorithm.
+        save_embed_format : str
+            Specify the format of saved embeddings.
+        """
+
+        device = self.device
+        # deal with uninitialized case first
+        if use_mini_batch_infer and \
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:74]: mini_batch should also work for edge class and edge regression.

File: python/graphstorm/inference/emb_infer.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+                inter_embs = node_mini_batch_gnn_predict(self._model, loader, return_proba,
+                                                  return_label=False)[1]
+                inter_embs = {ntype: inter_embs[ntype]} if isinstance(inter_embs, dict) \
+                    else {ntype: inter_embs}
+                g = loader.data.g
+                ntype_emb = create_dist_tensor((g.num_nodes(ntype), inter_embs[ntype].shape[1]),
+                                               dtype=inter_embs[ntype].dtype,
+                                               name=f'gen-emb-{ntype}',
+                                               part_policy=g.get_node_partition_policy(ntype),
+                                               persistent=True)
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:113]: We do not need persistent = True.

File: python/graphstorm/inference/emb_infer.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            # Currently it is not allowed to do mini-batch inference
+            # and save embedding on edge tasks
+            embs = do_full_graph_inference(self._model, loader.data, fanout=loader.fanout,
+                                           task_tracker=self.task_tracker)
+            target_ntypes = set()
+            for etype in loader.data.eval_etypes:
+                target_ntypes.add(etype[0])
+                target_ntypes.add(etype[2])
+
+            embs = {ntype: embs[ntype] for ntype in sorted(target_ntypes)}
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:131]: For mini-batch inference, we just iterate all target_ntypes and compute node embeddings for each ntype.

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    tracker = gs.create_builtin_task_tracker(config)
+    if gs.get_rank() == 0:
+        tracker.log_params(config.__dict__)
+
+    if config.task_type == BUILTIN_TASK_LINK_PREDICTION:
+        input_graph = GSgnnEdgeInferData(config.graph_name,
+                                        config.part_config,
+                                        eval_etypes=config.eval_etype,
+                                        node_feat_field=config.node_feat_name,
+                                        decoder_edge_feat=config.decoder_edge_feat)
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:56]: We do not need decoder_edge_feat.

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    elif config.task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
+        input_graph = GSgnnNodeInferData(config.graph_name,
+                                    config.part_config,
+                                    eval_ntypes=config.target_ntype,
+                                    node_feat_field=config.node_feat_name,
+                                    label_field=config.label_field)
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:62]: We do not need label_field

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    elif config.task_type in {BUILTIN_TASK_EDGE_CLASSIFICATION, BUILTIN_TASK_EDGE_REGRESSION}:
+        input_graph = GSgnnEdgeInferData(config.graph_name,
+                                    config.part_config,
+                                    eval_etypes=config.target_etype,
+                                    node_feat_field=config.node_feat_name,
+                                    label_field=config.label_field,
+                                    decoder_edge_feat=config.decoder_edge_feat)
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:69]: We do not need decoder_edge_feat

File: python/graphstorm/inference/emb_infer.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+    Inferrer wrapper for embedding generation.
+"""
+import logging
+from graphstorm.config import  (BUILTIN_TASK_NODE_CLASSIFICATION,
+                                BUILTIN_TASK_NODE_REGRESSION,
+                                BUILTIN_TASK_EDGE_CLASSIFICATION,
+                                BUILTIN_TASK_EDGE_REGRESSION,
+                                BUILTIN_TASK_LINK_PREDICTION)
+from .graphstorm_infer import GSInferrer
```
[the remark this answers is not in the document]
Comment (jalencato) [reply python/graphstorm/inference/emb_infer.py:24]: Here should be the graphstorm_infer.py file

File: python/graphstorm/inference/emb_infer.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+    to generate embedding for inferer.
+
+    Parameters
+    ----------
+    model : GSgnnNodeModel
+        The GNN model with different task.
+    """
+
+    def nc_emb(self, g, model, use_mini_batch_infer, fanout):
+        """ Embedding Generation for node task.
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:44]:
```suggestion
        """ Generate embeddings for node task.
```

File: python/graphstorm/inference/emb_infer.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+
+        device = self.device
+
+        assert save_embed_path is not None, \
+            "It requires save embed path for gs_gen_node_embedding"
+
+        sys_tracker.check('start embedding generation')
+        self._model.eval()
+
+        if task_type == BUILTIN_TASK_LINK_PREDICTION:
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:159]: Code is straightforward here. We just need to do:
for different task types, we get ntypes to do inference
then do inference.

So we do not need to implement lp_emb, nc_emb, etc.

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    gs.initialize(ip_config=config.ip_config, backend=config.backend,
+                  use_wholegraph=use_wholegraph(config.part_config))
+    rt_profiler.init(config.profile_path, rank=gs.get_rank())
+    sys_tracker.init(config.verbose, rank=gs.get_rank())
+    device = setup_device(config.local_rank)
+    tracker = gs.create_builtin_task_tracker(config)
+    if gs.get_rank() == 0:
+        tracker.log_params(config.__dict__)
+
+    if config.task_type == BUILTIN_TASK_LINK_PREDICTION:
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:46]: Do we need to init GSgnn Data here?
What we really need is actually config.eval_etype and config.eval_ntype, etc.

File: python/graphstorm/inference/emb_infer.py [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+        if use_mini_batch_infer:
+            embs = do_mini_batch_inference(self._model, data, fanout=eval_fanout,
+                                           edge_mask=None,
+                                           task_tracker=self.task_tracker,
+                                           infer_ntypes=infer_ntypes)
+        else:
+            embs = do_full_graph_inference(self._model, data, fanout=eval_fanout,
+                                           edge_mask=None,
+                                           task_tracker=self.task_tracker)
```
Comment (classicsong) [inline python/graphstorm/inference/emb_infer.py:95]: if infer_ntypes is not None, we only save embs in infer_ntypes, even it is full graph inference.

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        input_graph = GSgnnEdgeInferData(config.graph_name,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:47]:
```suggestion
        infer_data = GSgnnEdgeInferData(config.graph_name,
```

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+                                        config.part_config,
+                                        eval_etypes=config.eval_etype,
+                                        node_feat_field=config.node_feat_name)
+    elif config.task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
+        input_graph = GSgnnNodeInferData(config.graph_name,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:52]: ditto

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+                                        config.part_config,
+                                        eval_ntypes=config.target_ntype,
+                                        node_feat_field=config.node_feat_name)
+    elif config.task_type in {BUILTIN_TASK_EDGE_CLASSIFICATION, BUILTIN_TASK_EDGE_REGRESSION}:
+        input_graph = GSgnnEdgeInferData(config.graph_name,
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:57]: ditto

**Review by classicsong — APPROVED** [review]

**Pull request merged**